File: hdf-cli/pkg/diff/matching/fuzzy_match_test.go
package matching

import (
	"testing"

	hdf "github.com/mitre/hdf-cli/pkg/hdf"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// --- Tokenize unit tests ---

func TestTokenize_LowercaseAndSplit(t *testing.T) {
	tokens := Tokenize("Ensure SSH Root Login")
	assert.Contains(t, tokens, "ensure")
	assert.Contains(t, tokens, "ssh")
	assert.Contains(t, tokens, "root")
	assert.Contains(t, tokens, "login")
}

func TestTokenize_SplitOnPunctuation(t *testing.T) {
	tokens := Tokenize("root-login/access")
	assert.Contains(t, tokens, "root")
	assert.Contains(t, tokens, "login")
	assert.Contains(t, tokens, "access")
}

func TestTokenize_RemoveStopWords(t *testing.T) {
	tokens := Tokenize("the quick brown fox is a test")
	assert.NotContains(t, tokens, "the")
	assert.NotContains(t, tokens, "is")
	assert.NotContains(t, tokens, "a")
	assert.Contains(t, tokens, "quick")
	assert.Contains(t, tokens, "brown")
	assert.Contains(t, tokens, "fox")
}

func TestTokenize_EmptyString(t *testing.T) {
	tokens := Tokenize("")
	assert.Len(t, tokens, 0)
}

func TestTokenize_OnlyStopWords(t *testing.T) {
	tokens := Tokenize("the a an is")
	assert.Len(t, tokens, 0)
}

func TestTokenize_Deduplication(t *testing.T) {
	tokens := Tokenize("ssh ssh ssh login")
	// Set semantics: unique tokens only
	count := 0
	for _, tok := range tokens {
		if tok == "ssh" {
			count++
		}
	}
	assert.Equal(t, 1, count)
}

func TestTokenize_FilterShortTokens(t *testing.T) {
	// Tokens <= 2 chars should be filtered (except stop words are already filtered)
	// The TS implementation filters on length > 0 and stop words
	// The Go implementation should also filter tokens with length <= 2
	tokens := Tokenize("ab cd ensure")
	assert.NotContains(t, tokens, "ab")
	assert.NotContains(t, tokens, "cd")
	assert.Contains(t, tokens, "ensure")
}

// --- JaccardSimilarity unit tests ---

func TestJaccardSimilarity_IdenticalSets(t *testing.T) {
	a := mapSet("ssh", "root", "login")
	b := mapSet("ssh", "root", "login")
	assert.Equal(t, 1.0, JaccardSimilarity(a, b))
}

func TestJaccardSimilarity_DisjointSets(t *testing.T) {
	a := mapSet("ssh", "root", "login")
	b := mapSet("ntp", "time", "sync")
	assert.Equal(t, 0.0, JaccardSimilarity(a, b))
}

func TestJaccardSimilarity_PartialOverlap(t *testing.T) {
	a := mapSet("ssh", "root", "login", "disabled")
	b := mapSet("ssh", "root", "login", "must", "disabled")
	// intersection = {ssh, root, login, disabled} = 4
	// union = {ssh, root, login, disabled, must} = 5
	assert.InDelta(t, 0.8, JaccardSimilarity(a, b), 0.00001)
}

func TestJaccardSimilarity_BothEmpty(t *testing.T) {
	assert.Equal(t, 0.0, JaccardSimilarity(mapSet(), mapSet()))
}

func TestJaccardSimilarity_OneEmpty(t *testing.T) {
	assert.Equal(t, 0.0, JaccardSimilarity(mapSet("a"), mapSet()))
}

// --- FuzzyTitle strategy tests ---

func TestFuzzyTitleStrategy_Name(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	assert.Equal(t, "fuzzyTitle", s.Name())
}

func TestFuzzyTitleStrategy_MatchSimilarTitles(t *testing.T) {
	s := NewFuzzyTitleStrategy(0) // default threshold
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Title: strPtr("Ensure SSH root login is disabled"), Impact: 0.7},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Title: strPtr("SSH root login must be disabled"), Impact: 0.7},
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 1)
	assert.Equal(t, "V-001", result.Matched[0].OldReq.ID)
	assert.Equal(t, "RHEL-001", result.Matched[0].NewReq.ID)
	assert.GreaterOrEqual(t, result.Matched[0].Confidence, 0.6)
	assert.Equal(t, "fuzzyTitle", result.Matched[0].Strategy)
}

func TestFuzzyTitleStrategy_DissimilarTitles(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Title: strPtr("Ensure SSH root login is disabled"), Impact: 0.7},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Title: strPtr("Configure NTP time synchronization"), Impact: 0.7},
	}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

func TestFuzzyTitleStrategy_CustomThreshold(t *testing.T) {
	s := NewFuzzyTitleStrategy(0.9)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Title: strPtr("Ensure SSH root login is disabled"), Impact: 0.7},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Title: strPtr("SSH root login must be disabled"), Impact: 0.7},
	}

	result := s.Match(oldReqs, newReqs)

	// With 0.9 threshold, slightly different titles should not match
	assert.Len(t, result.Matched, 0)
}

func TestFuzzyTitleStrategy_GreedyBestMatch(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Title: strPtr("Ensure SSH root login is disabled"), Impact: 0.7},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Title: strPtr("SSH root login must be disabled"), Impact: 0.7},
		{ID: "RHEL-002", Title: strPtr("SSH protocol version must be 2"), Impact: 0.5},
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 1)
	assert.Equal(t, "RHEL-001", result.Matched[0].NewReq.ID)
	assert.Len(t, result.UnmatchedNew, 1)
}

func TestFuzzyTitleStrategy_NoTitle(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{{ID: "V-001", Impact: 0.7}}
	newReqs := []hdf.EvaluatedRequirement{{ID: "RHEL-001", Impact: 0.7}}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

func TestFuzzyTitleStrategy_EmptyTitles(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Title: strPtr(""), Impact: 0.7},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Title: strPtr(""), Impact: 0.7},
	}

	result := s.Match(oldReqs, newReqs)

	// Empty titles produce empty token sets -> similarity 0
	assert.Len(t, result.Matched, 0)
}

func TestFuzzyTitleStrategy_IdenticalTitles(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Title: strPtr("Ensure SSH root login is disabled"), Impact: 0.7},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Title: strPtr("Ensure SSH root login is disabled"), Impact: 0.7},
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 1)
	assert.Equal(t, 1.0, result.Matched[0].Confidence)
}

func TestFuzzyTitleStrategy_GreedySkipsAlreadyMatched(t *testing.T) {
	s := NewFuzzyTitleStrategy(0.3) // low threshold
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "1", Title: strPtr("SSH root login configuration check"), Impact: 0.7},
		{ID: "2", Title: strPtr("SSH root login setting verification"), Impact: 0.7},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "A", Title: strPtr("SSH root login configuration check"), Impact: 0.7},
	}

	result := s.Match(oldReqs, newReqs)

	// One match (best scoring pair), one unmatched old, zero unmatched new
	assert.Len(t, result.Matched, 1)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 0)
}

func TestFuzzyTitleStrategy_MultipleGreedyMatches(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Title: strPtr("Ensure SSH root login is disabled"), Impact: 0.7},
		{ID: "V-002", Title: strPtr("NTP time synchronization configured correctly"), Impact: 0.5},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Title: strPtr("SSH root login must be disabled"), Impact: 0.7},
		{ID: "RHEL-002", Title: strPtr("NTP time synchronization configured properly"), Impact: 0.5},
	}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 2)
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)
}

func TestFuzzyTitleStrategy_BothEmpty(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	result := s.Match(nil, nil)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)
}

func TestFuzzyTitleStrategy_NilTitle(t *testing.T) {
	s := NewFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{{ID: "V-001", Impact: 0.7, Title: nil}}
	newReqs := []hdf.EvaluatedRequirement{{ID: "RHEL-001", Impact: 0.7, Title: nil}}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

// --- JaccardSimilarity: a larger than b (swap branch) ---

func TestJaccardSimilarity_ALargerThanB(t *testing.T) {
	// a has more elements, b is smaller -- ensures the swap optimization is exercised
	a := mapSet("ssh", "root", "login", "disabled", "config", "setting")
	b := mapSet("ssh", "root")
	// intersection = {ssh, root} = 2
	// union = {ssh, root, login, disabled, config, setting} = 6
	expected := 2.0 / 6.0
	assert.InDelta(t, expected, JaccardSimilarity(a, b), 0.00001)
}

func TestJaccardSimilarity_BLargerThanA(t *testing.T) {
	a := mapSet("ssh", "root")
	b := mapSet("ssh", "root", "login", "disabled", "config", "setting")
	expected := 2.0 / 6.0
	assert.InDelta(t, expected, JaccardSimilarity(a, b), 0.00001)
}

func TestJaccardSimilarity_SingleElement(t *testing.T) {
	a := mapSet("ssh")
	b := mapSet("ssh")
	assert.Equal(t, 1.0, JaccardSimilarity(a, b))
}

// helper to create a set (as a map[string]bool for Jaccard tests).
func mapSet(items ...string) map[string]bool {
	s := make(map[string]bool, len(items))
	for _, item := range items {
		s[item] = true
	}
	return s
}

File: hdf-diff/test/matching/fuzzy-match.test.ts
import { describe, it, expect } from 'vitest';
import { createFuzzyTitleStrategy, tokenize, jaccardSimilarity } from '../../src/matching/fuzzy-match.js';

describe('FuzzyTitleStrategy', () => {
  describe('tokenize', () => {
    it('should lowercase and split on whitespace', () => {
      const tokens = tokenize('Ensure SSH Root Login');
      expect(tokens).toContain('ensure');
      expect(tokens).toContain('ssh');
      expect(tokens).toContain('root');
      expect(tokens).toContain('login');
    });

    it('should split on punctuation', () => {
      const tokens = tokenize('root-login/access');
      expect(tokens).toContain('root');
      expect(tokens).toContain('login');
      expect(tokens).toContain('access');
    });

    it('should remove common stop words', () => {
      const tokens = tokenize('the quick brown fox is a test');
      expect(tokens).not.toContain('the');
      expect(tokens).not.toContain('is');
      expect(tokens).not.toContain('a');
      expect(tokens).toContain('quick');
      expect(tokens).toContain('brown');
      expect(tokens).toContain('fox');
    });

    it('should return empty set for empty string', () => {
      const tokens = tokenize('');
      expect(tokens.size).toBe(0);
    });

    it('should return empty set for only stop words', () => {
      const tokens = tokenize('the a an is');
      expect(tokens.size).toBe(0);
    });
  });

  describe('jaccardSimilarity', () => {
    it('should return 1.0 for identical sets', () => {
      const a = new Set(['ssh', 'root', 'login']);
      const b = new Set(['ssh', 'root', 'login']);
      expect(jaccardSimilarity(a, b)).toBe(1.0);
    });

    it('should return 0.0 for completely disjoint sets', () => {
      const a = new Set(['ssh', 'root', 'login']);
      const b = new Set(['ntp', 'time', 'sync']);
      expect(jaccardSimilarity(a, b)).toBe(0.0);
    });

    it('should return correct similarity for partially overlapping sets', () => {
      const a = new Set(['ssh', 'root', 'login', 'disabled']);
      const b = new Set(['ssh', 'root', 'login', 'must', 'disabled']);
      // intersection = {ssh, root, login, disabled} = 4
      // union = {ssh, root, login, disabled, must} = 5
      expect(jaccardSimilarity(a, b)).toBeCloseTo(0.8, 5);
    });

    it('should return 0.0 for two empty sets', () => {
      expect(jaccardSimilarity(new Set(), new Set())).toBe(0.0);
    });

    it('should return 0.0 when one set is empty', () => {
      expect(jaccardSimilarity(new Set(['a']), new Set())).toBe(0.0);
    });
  });

  describe('strategy', () => {
    const strategy = createFuzzyTitleStrategy();

    it('should have name "fuzzyTitle"', () => {
      expect(strategy.name).toBe('fuzzyTitle');
    });

    it('should match similar titles above default threshold', () => {
      const oldReqs = [
        { id: 'V-001', title: 'Ensure SSH root login is disabled', impact: 0.7 },
      ];
      const newReqs = [
        { id: 'RHEL-001', title: 'SSH root login must be disabled', impact: 0.7 },
      ];

      const result = strategy.match(oldReqs, newReqs);

      expect(result.matched).toHaveLength(1);
      expect(result.matched[0]!.oldReq['id']).toBe('V-001');
      expect(result.matched[0]!.newReq['id']).toBe('RHEL-001');
      expect(result.matched[0]!.confidence).toBeGreaterThanOrEqual(0.6);
      expect(result.matched[0]!.strategy).toBe('fuzzyTitle');
    });

    it('should not match dissimilar titles', () => {
      const oldReqs = [
        { id: 'V-001', title: 'Ensure SSH root login is disabled', impact: 0.7 },
      ];
      const newReqs = [
        { id: 'RHEL-001', title: 'Configure NTP time synchronization', impact: 0.7 },
      ];

      const result = strategy.match(oldReqs, newReqs);

      expect(result.matched).toHaveLength(0);
      expect(result.unmatchedOld).toHaveLength(1);
      expect(result.unmatchedNew).toHaveLength(1);
    });

    it('should use custom minConfidence threshold', () => {
      const highThreshold = createFuzzyTitleStrategy(0.9);

      const oldReqs = [
        { id: 'V-001', title: 'Ensure SSH root login is disabled', impact: 0.7 },
      ];
      const newReqs = [
        { id: 'RHEL-001', title: 'SSH root login must be disabled', impact: 0.7 },
      ];

      const result = highThreshold.match(oldReqs, newReqs);

      // With 0.9 threshold, slightly different titles should not match
      expect(result.matched).toHaveLength(0);
    });

    it('should use greedy best-match (prefer highest similarity)', () => {
      const oldReqs = [
        { id: 'V-001', title: 'Ensure SSH root login is disabled', impact: 0.7 },
      ];
      const newReqs = [
        { id: 'RHEL-001', title: 'SSH root login must be disabled', impact: 0.7 },
        { id: 'RHEL-002', title: 'SSH protocol version must be 2', impact: 0.5 },
      ];

      const result = strategy.match(oldReqs, newReqs);

      expect(result.matched).toHaveLength(1);
      expect(result.matched[0]!.newReq['id']).toBe('RHEL-001');
      expect(result.unmatchedNew).toHaveLength(1);
    });

    it('should handle requirements without title field', () => {
      const oldReqs = [{ id: 'V-001', impact: 0.7 }];
      const newReqs = [{ id: 'RHEL-001', impact: 0.7 }];

      const result = strategy.match(oldReqs, newReqs);

      expect(result.matched).toHaveLength(0);
      expect(result.unmatchedOld).toHaveLength(1);
      expect(result.unmatchedNew).toHaveLength(1);
    });

    it('should handle empty title strings', () => {
      const oldReqs = [{ id: 'V-001', title: '', impact: 0.7 }];
      const newReqs = [{ id: 'RHEL-001', title: '', impact: 0.7 }];

      const result = strategy.match(oldReqs, newReqs);

      // Empty titles produce empty token sets -> similarity 0
      expect(result.matched).toHaveLength(0);
    });

    it('should match identical titles with confidence 1.0', () => {
      const oldReqs = [
        { id: 'V-001', title: 'Ensure SSH root login is disabled', impact: 0.7 },
      ];
      const newReqs = [
        { id: 'RHEL-001', title: 'Ensure SSH root login is disabled', impact: 0.7 },
      ];

      const result = strategy.match(oldReqs, newReqs);

      expect(result.matched).toHaveLength(1);
      expect(result.matched[0]!.confidence).toBe(1.0);
    });

    it('should skip already-matched new requirements in greedy assignment', () => {
      const strategy = createFuzzyTitleStrategy(0.3); // low threshold so both old reqs match
      const oldReqs = [
        { id: '1', title: 'SSH root login configuration check', impact: 0.7 },
        { id: '2', title: 'SSH root login setting verification', impact: 0.7 },
      ];
      const newReqs = [
        // Both old reqs are similar to this single new req;
        // the greedy algorithm matches the best one and skips the other
        { id: 'A', title: 'SSH root login configuration check', impact: 0.7 },
      ];
      const result = strategy.match(
        oldReqs as unknown as Record<string, unknown>[],
        newReqs as unknown as Record<string, unknown>[],
      );
      // One match (best scoring pair), one unmatched old, zero unmatched new
      expect(result.matched).toHaveLength(1);
      expect(result.unmatchedOld).toHaveLength(1);
      expect(result.unmatchedNew).toHaveLength(0);
    });

    it('should match multiple requirements greedily', () => {
      const oldReqs = [
        { id: 'V-001', title: 'Ensure SSH root login is disabled', impact: 0.7 },
        { id: 'V-002', title: 'NTP time synchronization configured correctly', impact: 0.5 },
      ];
      const newReqs = [
        { id: 'RHEL-001', title: 'SSH root login must be disabled', impact: 0.7 },
        { id: 'RHEL-002', title: 'NTP time synchronization configured properly', impact: 0.5 },
      ];

      const result = strategy.match(oldReqs, newReqs);

      expect(result.matched).toHaveLength(2);
      expect(result.unmatchedOld).toHaveLength(0);
      expect(result.unmatchedNew).toHaveLength(0);
    });
  });
});
